src/python/pants/engine/build_files.py

```diff
@@ -135,7 +135,9 @@ def collect_dependencies(item):
         maybe_append(key, value)
 
   collect_dependencies(struct)
-  return UnhydratedStruct(address, struct, dependencies)
+
+  return UnhydratedStruct(
+    filter(lambda build_address: build_address == address, addresses)[0], struct, dependencies)
```

Comment:
is there always guaranteed to be at least one non-filtered address here? if not, the index access here may be unsafe.